### PR TITLE
Fix EC_KEY_set_private_key() NULL private key behavior in 1.1.1

### LINKED ISSUE #18744: Regression: Unable to set EC_KEY private_key to NULL starting in 1.1.1h
OpenSSL 1.1.1h introduces a behavior change wherein one can no longer set an EC_KEY's private_key to NULL.

This behavior changes in 6a01f6f4b41d045e2a3abcb10163633d769db76a. Based on the original PR (https://github.com/openssl/openssl/pull/11127), it appears that this is hardening backported from 3.0.0 that unintentionally introduced a regression.

From https://github.com/openssl/openssl/commit/6a01f6f4b41d045e2a3abcb10163633d769db76a#diff-38a47bf175cde97de2df18870a9ab300366f191a848bdea84def068fe0fde69fR484-R486: 

```diff
+    tmp_key = BN_dup(priv_key);
+    if (tmp_key == NULL)
+        return 0;
```

`EC_KEY_set_private_key` returns early and never sets `key->priv_key` to NULL.

### PR DESCRIPTION
**Pull request by robertohueso**
This allows to set EC_KEY's private key to NULL.

Fixes #18744 in 1.1.1 branch.

### CODE REVIEW

**Comment by romen** [comment]
I raised a question for OTC on what is the intended behavior on the original issue: https://github.com/openssl/openssl/issues/18744#issuecomment-1195068699

File: crypto/ec/ec_key.c
Comment (t8m) [inline crypto/ec/ec_key.c:486]: The old priv_key needs to be freed

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/ec/ec_key.c
Comment (romen) [reply crypto/ec/ec_key.c:486]: I would actually recommend to keep L485-L488 as they were, and add a new `if` block around L445, something similar to this:

~~~c
if (priv_key == NULL) {
    /**
     * If passed NULL, we properly clear `key->priv_key`.
     * We return `0` to comply with legacy behavior for this function,
     * see <https://github.com/openssl/openssl/issues/18744#issuecomment-1195175696>.
     */
    BN_clear_free(key->priv_key);
    key->priv_key = NULL;
    return 0; /* intentional for legacy compatibility */
}
~~~

**Review by romen — CHANGES_REQUESTED** [review]
Thanks @robertohueso for this PR and for reporting the original issue.

After the discussion on #18744 , I believe we can alter your proposal to include @t8m reminder that we need to securely clear and free the private key already stored within the `EC_KEY` object, and also document that we have a special case when the caller pass `NULL` as the private key, as we clear the inner field but still return an error (weird, but we do this for compatibility with legacy behavior).

The same approach should be taken for `3.0` and `master`, where this function is deprecated so we don't want to alter its behavior compared to `1.1.1`.

@robertohueso would you also like to take the challenge of adding a regression test for this behavior and improve the documentation for the public function?

**Comment by robertohueso** [comment]
Thank you @romen and @t8m for your comments! Working on the improvements (included tests and docs) :construction_worker:

**Review by romen — CHANGES_REQUESTED** [review]
Thanks @robertohueso for addressing the feedback: progress looks very good to me.

From my point of view the only thing left is editing `doc/man3/EC_KEY_new.pod` to include a mention of `EC_KEY_set_private_key` accepting `NULL` as the `prv` argument to securely clear the _private key_ component from the given `EC_KEY` object, in the `DESCRIPTION` section, and to add a note in the `RETURN VALUES` section, clarifying that when passing `prv == NULL` to this function we always return `0`, for legacy compatibility, and that such case should not be treated as an error.

**Comment by robertohueso** [comment]
@romen I think this is ready for review :)

File: doc/man3/EC_KEY_new.pod
Comment (romen) [inline doc/man3/EC_KEY_new.pod:106]: In the SYNOPSIS section the argument is called `prv`, we should use the same name in all sections.

Personally, I am OK with editing the line in SYNOPSIS to call it `priv_key` instead, if you believe it improves readability.
Alternatively DESCRIPTION and RETURN VALUES edits should use `prv`.

@t8m do you have a preference on either approach?

Comment (romen) [reply doc/man3/EC_KEY_new.pod:106]: Also @t8m might help me remembering the appropriate formatting "tag" for argument names and function names: on that regard it seems the whole page needs reformatting so maybe we can defer that to a separate dedicated reformatting PR?

**Review by romen — COMMENTED** [review]
Looking good, thanks! Last nits!

File: doc/man3/EC_KEY_new.pod
Comment (robertohueso) [reply doc/man3/EC_KEY_new.pod:106]: Just fixed the `priv_key` issue, thanks!

**Review by romen — APPROVED** [review]
LGTM!

Thanks @robertohueso!

**Comment by romen** [comment]
@robertohueso we now have to wait for another review. In the meantime you might want to check locally if this cherry-picks cleanly to 3.0-stable/master. If that is not the case we will need dedicated PRs.

**Review by mattcaswell — APPROVED** [review]

**Comment by tmshort** [comment]
Removing 3.0 and master tags due to #18941 and #18942

**Comment by romen** [comment]
Thanks!

Merged:
- #18942 to `master` as:
   - b304b3e8f7 Fix EC_KEY_set_private_key() priv_key regression
   - d93f154d5a Add test for EC_KEY_set_private_key()
   - fae06b5779 Fix EC_KEY_set_private_key() NULL priv_key docs
- #18941 to `openssl-3.0` as:
   - f76cf4d988 Fix EC_KEY_set_private_key() priv_key regression
   - 66627cdfc2 Add test for EC_KEY_set_private_key()
   - c682dfdbdf Fix EC_KEY_set_private_key() NULL priv_key docs
- #18874 to `OpenSSL_1_1_1-stable` as:
   - 143d7d4c79 Fix EC_KEY_set_private_key() priv_key regression
   - 1c2f52bed3 Add test for EC_KEY_set_private_key()
   - 4648839300 Fix EC_KEY_set_private_key() NULL priv_key docs

**Pull request closed without merging**
